Sources/Testing/SourceAttribution/Expression.swift

```diff
@@ -0,0 +1,261 @@
+    /// ``Expression/Kind/functionCall`` case.
+    ///
+    /// This type is not part of the public interface of the testing library.
+    struct FunctionCallArgument: Sendable {
+      /// The label, if present, of the argument.
+      var label: String?
+
+      /// The value, as an expression, of the argument.
+      var value: Expression
+    }
+
+    /// The expression represents a function call.
+    ///
+    /// - Parameters:
+    ///   - value: The value on which the function was called, if any.
+    ///   - functionName: The name of the function that was called.
+    ///   - arguments: The arguments passed to the function.
+    indirect case functionCall(value: Expression?, functionName: String, arguments: [FunctionCallArgument])
+
+    /// The expression represets a property access.
+    ///
+    /// - Parameters:
+    ///   - value: The value whose property was accessed.
+    ///   - keyPath: The key path, relative to `value`, that was accessed, not
+    ///     including a leading backslash or period.
+    indirect case propertyAccess(value: Expression, keyPath: Expression)
+  }
+
+  /// The kind of syntax node represented by this instance.
+  ///
+  /// This property is not part of the public interface of the testing library.
+  /// Use `String(describing:)` to access the source code represented by an
+  /// instance of this type.
+  var kind: Kind
+
+  /// The source code of the original captured expression.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var sourceCode: String {
+    switch kind {
+    case let .generic(sourceCode):
+      return sourceCode
+    case let .binaryOperation(lhs, op, rhs):
+      return "\(lhs) \(op) \(rhs)"
+    case let .functionCall(value, functionName, arguments):
+      let argumentList = arguments.lazy
+        .map { argument in
+          if let argumentLabel = argument.label {
+            return "\(argumentLabel): \(argument.value.sourceCode)"
+          }
+          return argument.value.sourceCode
+        }.joined(separator: ", ")
+      if let value {
+        return "\(value.sourceCode).\(functionName)(\(argumentList))"
+      }
+      return "\(functionName)(\(argumentList))"
+    case let .propertyAccess(value, keyPath):
+      return "\(value.sourceCode).\(keyPath.sourceCode)"
+    }
+  }
+
+  /// A description, captured using ``String/init(describingForTest:)``, of
+  /// the runtime value of this subexpression.
+  ///
+  /// If the runtime value of this subexpression has not been evaluated, the
+  /// value of this property is `nil`.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var runtimeValueDescription: String?
+
+  /// The fully-qualified name of the type of this subexpression's runtime
+  /// value.
+  ///
+  /// If the runtime value of this subexpression has not been evaluated, the
+  /// value of this property is `nil`.
+  @_spi(ExperimentalSourceCodeCapturing)
+  public var fullyQualifiedTypeNameOfRuntimeValue: String?
```

Comment:
If there is a need to decompose this to (for example) omit showing the module name, would you suggest splitting on period character? Is there any benefit to exposing a tuple or other structured representation of this?

Reply:
Yes, you'd split on the first period character. We don't try to represent type names as a tuple/structure elsewhere, so I wasn't planning to do so here.